File: backend/controller/menu.py
# router service db router+service db
from fastapi import APIRouter

from core.utils import list_to_tree
from dbhelper.menu import (del_menu, get_menu, get_tree_menu, insert_menu,
                           put_menu)
from schemas import MenuIn, MenuRead, Response

router = APIRouter(prefix="/menu", tags=["菜单管理"])


@router.post("", summary="菜单新增", response_model=Response[MenuRead])
async def menu_add(data: MenuIn):
    return Response(data=await insert_menu(data))


@router.get("", summary="菜单列表", response_model=Response)
async def menu_arr():
    menus = await get_tree_menu()
    try:
        data = list_to_tree(menus)
    except KeyError:
        return Response(code=400, msg="菜单根节点丢失")
    return Response(data=data)


@router.delete("/{pk}", summary="菜单删除", response_model=Response)
async def menu_del(pk: int):
    if await get_menu({"pid": pk}) is not None:
        return Response(code=400, msg="请先删除子节点")
    if await del_menu(pk) == 0:
        return Response(code=400, msg="菜单不存在")
    return Response()


@router.put("/{pk}", summary="菜单更新", response_model=Response)
async def menu_put(pk: int, data: MenuIn):
    """更新菜单"""
    if await put_menu(pk, data) == 0:
        return Response(code=400, msg="菜单不存在")
    return Response()

File: frontend/src/service/menu.js
import request from '@/utils/request'

// 获取角色列表, 需要考虑创建用户选择角色应该是所有未被删除的情况
export function getMenus(params) {
  return request({
    url: '/menu',
    params
  })
}

// 条件查询
export function queryMenu(data) {
  return request({
    url: '/menu/query',
    method: 'post',
    data
  })
}

// 删除
export function delMenu(id) {
  return request({
    url: `/menu/${id}`,
    method: 'delete'
  })
}

// 创建
export function addMenu(data) {
  return request({
    url: '/menu',
    method: 'post',
    data
  })
}

// 修改
export function putMenu(id, data) {
  return request({
    url: `/menu/${id}`,
    method: 'put',
    data
  })
}
